src/parser/parser.lisp

```diff
@@ -0,0 +1,1491 @@
+  (declares  nil                      :type toplevel-declare-list         :read-only nil)
+  (defines   nil                      :type toplevel-define-list          :read-only nil)
+  (classes   nil                      :type toplevel-define-class-list    :read-only nil)
+  (instances nil                      :type toplevel-define-instance-list :read-only nil))
+
+;;
+;; Empty package for reading (package) forms
+;;
+(defpackage #:coalton-impl/parser/read
+  (:use))
+
+;; TODO: handle trailing attributes
+
+(defun parse-file (filename)
+  (with-open-file (file filename :if-does-not-exist :error)
+    (let* (;; Setup eclector readtable
+           (eclector.readtable:*readtable*
+             (eclector.readtable:copy-readtable eclector.readtable:*readtable*))
+
+           ;; Initial package to read (package) forms into
+           (*package* (find-package "COALTON-IMPL/PARSER/READ"))
+
+           ;; Read unspecified floats as double floats
+           (*read-default-float-format* 'double-float)
+
+           ;; Read the (package) form
+           (package-form (eclector.concrete-syntax-tree:read file nil 'eof)))
+
+      (when (eq package-form 'eof)
+        (error "file starts with eof"))
+
+      (setf *package* (parse-package package-form file))
+
+      (let* ((program (make-program :package *package* :file file))
+
+             (attributes (make-array 0 :adjustable t :fill-pointer t)))
+
+        (block parse-loop
+          (loop :with elem := nil
+                :do (setf elem (eclector.concrete-syntax-tree:read file nil 'eof))
+
+                :when (eq elem 'eof)
+                  :do (return-from parse-loop)
+
+                :do (when (and (parse-toplevel-form elem program attributes file)
+                               (plusp (length attributes)))
+                      (util:coalton-bug "parse-toplevel-form indicated that a form was parsed but did not
+consume all attributes"))))
+
+        (unless (zerop (length attributes))
+          (error 'parse-error
+                 :err (coalton-error
+                       (cdr (aref attributes 0)) file
+                       :message "Orphan attribute"
+                       :primary-note "attribute must be attached to another form")))
+
+        (setf (program-types program) (nreverse (program-types program)))
+        (setf (program-declares program) (nreverse (program-declares program)))
+        (setf (program-defines program) (nreverse (program-defines program)))
+        (setf (program-classes program) (nreverse (program-classes program)))
+
+        program))))
+
+(defun parse-package (form file)
+  "Parses a coalton package decleration in the form of (package {name})"
+  (declare (type cst:cst form)
+           (type sb-sys::fd-stream file)
+           (values package))
+
+  ;; Package declarations must start with "PACKAGE"
+  (unless (equalp (symbol-name (cst:raw (cst:first form))) "PACKAGE")
```

Comment:
`string=`

Reply:
Or `string-equalp`, depending on if you want case-sensitivity.